contracts/ForeignBridge.sol

```diff
@@ -2,74 +2,92 @@ pragma solidity 0.4.19;
-    event Withdraw(address recipient, uint value);
+    event Withdraw(address recipient, uint256 value, uint256 homeGasPrice);
 
     /// Collected signatures which should be relayed to home chain.
     event CollectedSignatures(address authorityResponsibleForRelay, bytes32 messageHash);
 
+    event TokenAddress(address token);
+
+    event GasConsumptionLimitsUpdated(uint256 gasLimitDepositRelay, uint256 gasLimitWithdrawConfirm);
+
+    event SignedForDeposit(address indexed signer, bytes32 message);
+    event SignedForWithdraw(address indexed signer, bytes32 message);
+
     function ForeignBridge(
-        address _token,
-        address _validatorsContract
-    ) public
-    {
-        require(_token != address(0) && _validatorsContract != address(0));
-        validatorsContract = BridgeValidators(_validatorsContract);
-        token = _token;
+        address _validatorContract,
+        address _erc677token
+    ) public Validatable(_validatorContract) {
+        erc677token = IBurnableMintableERC677Token(_erc677token);
     }
 
+    function setGasLimitDepositRelay(uint256 gas) onlyValidator {
+        gasLimitDepositRelay = gas;
 
-    /// Used to deposit money to the contract.
-    ///
-    /// deposit recipient (bytes20)
-    /// deposit value (uint)
-    /// mainnet transaction hash (bytes32) // to avoid transaction duplication
-    function deposit(address recipient, uint256 value, bytes32 transactionHash) public {
-        require(validatorsContract.onlyValidator(msg.sender));
+        GasConsumptionLimitsUpdated(gasLimitDepositRelay, gasLimitWithdrawConfirm);
+    }
+
+    function setGasLimitWithdrawConfirm(uint256 gas) onlyValidator {
+        gasLimitWithdrawConfirm = gas;
+
+        GasConsumptionLimitsUpdated(gasLimitDepositRelay, gasLimitWithdrawConfirm);
+    }
+
+    function deposit(address recipient, uint value, bytes32 transactionHash) public onlyValidator {
+        require(erc677token != address(0x0));
+
+        // Protection from misbehaing authority
         bytes32 hash_msg = keccak256(recipient, value, transactionHash);
         bytes32 hash_sender = keccak256(msg.sender, hash_msg);
 
-        // Prevents Duplicated deposits
+        // Duplicated deposits
         require(!deposits_signed[hash_sender]);
         deposits_signed[hash_sender] = true;
 
-        uint8 signed = num_deposits_signed[hash_msg] + 1;
+        uint signed = num_deposits_signed[hash_msg] + 1;
         num_deposits_signed[hash_msg] = signed;
 
-        if (signed == requiredValidators) {
-            // token.mint(recipient, value);
+        if (signed == validatorContract.requiredSignatures()) {
+            // If the bridge contract does not own enough tokens to transfer
+            // it will couse funds lock on the home side of the bridge
+            erc677token.mint(recipient, value);
             Deposit(recipient, value);
         }
+        SignedForDeposit(msg.sender, transactionHash);
```

Comment:
let's have:
```
       SignedForDeposit(msg.sender, transactionHash);
       if (signed == validatorContract.requiredSignatures()) {
           // If the bridge contract does not own enough tokens to transfer
           // it will couse funds lock on the home side of the bridge
           erc677token.mint(recipient, value);
           Deposit(recipient, value);
         }
```
It will be more logical: `SignedDeposit` prior to `Deposit`